File: leetcode/questions/680/java/680.java
class Solution {
    private boolean check(int leftIdx, int rightIdx, boolean previouslyDeleted, String s) {
        while (leftIdx < rightIdx) {
            if (s.charAt(leftIdx) != s.charAt(rightIdx)) {
                if (previouslyDeleted) {
                    return false;
                }
                // Try to delete one character
                return check(leftIdx + 1, rightIdx, true, s) || check(leftIdx, rightIdx - 1, true, s);
            }
            ++leftIdx;
            --rightIdx;
        }
        return true;
    }

    public boolean validPalindrome(String s) {
        return check(0, s.length() - 1, false, s);
    }
}

File: leetcode/questions/680/python/680.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        def check(left, right, previously_deleted):
            while left < right:
                if previously_deleted:
                    return False
                
                if s[left] != s[right]:
                    return check(left + 1, right, True) or check(left, right - 1, True)

                left += 1
                right -= 1

            return True

        return check(0, len(s) - 1, False)